src/irr/asset/material_compiler/CMaterialCompilerGLSLBackendCommon.cpp

```diff
@@ -0,0 +1,1687 @@
+			defs += "\n#define PREFETCH_TEX_3";
+	}
+
+	//parameter numbers
+	{
+		defs += "\n#define PARAMS_ALPHA_U_IX " + std::to_string(ALPHA_U_TEX_IX);
+		defs += "\n#define PARAMS_ALPHA_V_IX " + std::to_string(ALPHA_V_TEX_IX);
+		defs += "\n#define PARAMS_REFLECTANCE_IX " + std::to_string(REFLECTANCE_TEX_IX);
+		defs += "\n#define PARAMS_TRANSMITTANCE_IX " + std::to_string(TRANSMITTANCE_TEX_IX);
+		defs += "\n#define PARAMS_SIGMA_A_IX " + std::to_string(SIGMA_A_TEX_IX);
+		defs += "\n#define PARAMS_WEIGHT_IX " + std::to_string(WEIGHT_TEX_IX);
+		defs += "\n#define PARAMS_OPACITY_IX " + std::to_string(OPACITY_TEX_IX);
+	}
+
+	for (uint32_t i = 0u; i < INSTR_MAX_PARAMETER_COUNT; ++i)
+	{
+		using namespace std::string_literals;
+		if (_res.paramTexPresence[0][1] == 0u)
+			defs += "\n#define PARAM"s + static_cast<char>(i+1u+'0') + "_NEVER_TEX";
+		else if (_res.paramTexPresence[0][0] == 0u)
+			defs += "\n#define PARAM"s + static_cast<char>(i+1u+'0') + "_ALWAYS_TEX";
+
+		if (_res.paramConstants[i].first)
+		{
+			const auto& c = _res.paramConstants[i].second;
+			defs += "\n#define PARAM"s + static_cast<char>(i+1u+'0') + "_ALWAYS_SAME_VALUE";
+			defs += "\n#define PARAM"s + static_cast<char>(i+1u+'0') + "_VALUE vec3(" +
+				std::to_string(c.x) + ", " + std::to_string(c.y) + ", " + std::to_string(c.z) + ")";
+		}
+	}
+
+
+	defs += "\n";
+
+	return defs;
+}
+
+core::unordered_map<uint32_t, uint32_t> CMaterialCompilerGLSLBackendCommon::createBsdfDataIndexMapForPrefetchedTextures(SContext* _ctx, const instr_stream::traversal_t& _tex_prefetch_stream, const core::unordered_map<instr_stream::STextureData, uint32_t, instr_stream::STextureData::hash>& _tex2reg) const
+{
+	core::unordered_map<uint32_t, uint32_t> ix2ix;
+
+	for (instr_t instr : _tex_prefetch_stream)
+	{
+		const auto bsdf_ix = core::bitfieldExtract(instr, BITFIELDS_BSDF_BUF_OFFSET_SHIFT, BITFIELDS_BSDF_BUF_OFFSET_WIDTH);
+		{
+			auto found = ix2ix.find(bsdf_ix);
+			if (found != ix2ix.end())
+				continue;
+		}
+
+		const SBSDFUnion& bsdf_data = (*_ctx->pBsdfData)[bsdf_ix];
+		auto new_bsdf_data = bsdf_data;
+		constexpr uint32_t tex_fetch_flag_shift[tex_prefetch::BITFIELDS_FETCH_TEX_COUNT]{
+			tex_prefetch::BITFIELDS_FETCH_TEX_0_SHIFT,
+			tex_prefetch::BITFIELDS_FETCH_TEX_1_SHIFT,
+			tex_prefetch::BITFIELDS_FETCH_TEX_2_SHIFT,
+			tex_prefetch::BITFIELDS_FETCH_TEX_3_SHIFT
+		};
+		for (uint32_t i = 0u; i < SBSDFUnion::MAX_TEXTURES; ++i)
+			if (core::bitfieldExtract(instr, tex_fetch_flag_shift[i], 1))
+			{
+				auto found = _tex2reg.find(bsdf_data.param[i].tex);
+				if (found == _tex2reg.end())
+					continue;
+
+				new_bsdf_data.param[i].tex.prefetch_reg = found->second;
+			}
+
+		ix2ix.insert({ bsdf_ix, _ctx->pBsdfData->size() });
+		_ctx->pBsdfData->push_back(new_bsdf_data);
+	}
+
+	return ix2ix;
+}
+
+auto CMaterialCompilerGLSLBackendCommon::compile(SContext* _ctx, IR* _ir, bool _computeGenChoiceStream) -> result_t
```

Comment:
we need to take the `IR` as a `const` and make a copy, otherwise the temporary nodes created on a `compile` would eventually overrun the storage

Reply:
TODO

Reply:
STILL NOT CONST POINTER TO IR

Reply:
still TODO

Reply:
i did this the other way (freeing mem for temp nodes after compilation), you've seen

Reply:
eh ok, I still dont like it... but lets just compromise for the sake of the merge